src/tests/functional/page_refresh_tests.js

```diff
@@ -180,12 +180,18 @@ test("uses morphing to update remote frames marked with refresh='morph'", async
 test("don't refresh frames contained in [data-turbo-permanent] elements", async ({ page }) => {
   await page.goto("/src/tests/fixtures/page_refresh.html")
 
+  await page.evaluate(() => {
+    const frame = document.getElementById("remote-permanent-frame");
+    if (frame) {
+      frame.textContent = "Frame to be preserved"
+    }
+  }) // The final assertion cannot be noNextEventOnTarget because the assertion passes even when it reloads the frame.
+
```

Comment:
Could this assertion rely on Playwright to find the `#remote-permanent-frame` using its synchronization mechanisms instead of `document.getElementById` and a conditional?

```suggestion
  const frame = await page.locator("#remote-permanent-frame")
  
  // The final assertion cannot be noNextEventOnTarget because the assertion passes even when it reloads the frame.
  await frame.evaluate((frame) => frame.textContent = "Frame to be preserved")
```

Reply:
Good call, yes. I just tested and that successfully worked to catch the broken state. I committed this change.